### PR TITLE
16279 - Custom downloader

### PR DESCRIPTION
**Pull request by petrca**
Contains new parallel custom download manager to download large files faster. It uses http requests with "Range" header param to specify part size. Supports proxy also. It was developed quite fast and some parts of code could be later improved.

### CODE REVIEW

**Comment by petrca** [comment]
Stay tuned, more comments will be added.

File: src/Squirrel/UpdateManager.CheckForUpdates.cs [hunk 1]
```diff
@@ -24,9 +24,9 @@ public CheckForUpdateImpl(string rootAppDirectory)
                 UpdaterIntention intention,
                 string localReleaseFile,
                 string updateUrlOrPath,
+				string token,
```
Comment (ArosPrince) [inline src/Squirrel/UpdateManager.CheckForUpdates.cs:27]: Alignment.

File: src/Squirrel/UpdateManager.CheckForUpdates.cs [hunk 2]
```diff
@@ -62,33 +62,28 @@ public CheckForUpdateImpl(string rootAppDirectory)
...
-
-                        if (retries <= 0) throw;
-                        retries--;
-                        goto retry;
-                    }
-
+                    
+                    var uri = Utility.AppendPathToUri(new Uri(updateUrlOrPath), "RELEASES");
+
+					// Not working with custom queries
```
Comment (ArosPrince) [inline src/Squirrel/UpdateManager.CheckForUpdates.cs:68]: Please elaborate on this...

I can remember what's this about now. But would I after several months? Not likely...

File: src/Squirrel/UpdateManager.CheckForUpdates.cs [hunk 3]
```diff
@@ -62,33 +62,28 @@ public CheckForUpdateImpl(string rootAppDirectory)
...
+					/*if (latestLocalRelease != null) {
+                        uri = Utility.AddQueryParamsToUri(uri, new Dictionary<string, string> {
+                            { "id", latestLocalRelease.PackageName },
+                            { "localVersion", latestLocalRelease.Version.ToString() },
+                            { "arch", Environment.Is64BitOperatingSystem ? "amd64" : "x86" }
+                        });
+                    }*/
+
+					string tempFilePath = Path.GetTempFileName();
```
Comment (ArosPrince) [inline src/Squirrel/UpdateManager.CheckForUpdates.cs:77]: I'd just run autoformat on the whole file...

File: test/UpdateManagerTests.cs [hunk 4]
```diff
@@ -171,7 +171,9 @@ public async Task WhenBothFilesAreInSyncNoUpdatesAreApplied()
 
                     // check for an update
                     UpdateInfo updateInfo;
-                    using (var mgr = new UpdateManager(remotePackages, "theApp", tempDir, new FakeUrlDownloader())) {
+
+					// PO: TODO fix fake downloader test
```
Comment (ArosPrince) [inline test/UpdateManagerTests.cs:175]: ?

**Review by klnmi97 — APPROVED** [review]

File: src/Squirrel/UpdateManager.CheckForUpdates.cs (hunk 1 above)
Comment (petrca) [reply src/Squirrel/UpdateManager.CheckForUpdates.cs:27]: I fixed it in all changed files.

File: src/Squirrel/UpdateManager.CheckForUpdates.cs (hunk 3 above)
Comment (petrca) [reply src/Squirrel/UpdateManager.CheckForUpdates.cs:77]: Fixed.

File: test/UpdateManagerTests.cs (hunk 4 above)
Comment (petrca) [reply test/UpdateManagerTests.cs:175]: See redmine #16376.

File: src/Squirrel/UpdateManager.CheckForUpdates.cs (hunk 2 above)
Comment (petrca) [reply src/Squirrel/UpdateManager.CheckForUpdates.cs:68]: Updated.

**Pull request merged**